File: functions/src/caps.js
/**
 * Caps system — the most complex algorithmic component.
 *
 * Handles:
 * - Caps call verification (can the player guarantee all remaining rounds?)
 * - Late caps detection (server-side tracking of obligation)
 * - Wrong/Early caps detection
 * - External caps (from opposition)
 * - "Claim remaining rounds" convenience shortcut
 *
 * Algorithm:
 * For a given game state and player, enumerate all valid distributions of
 * unknown cards among opponents. For each distribution, simulate the
 * proposed play order against all possible opponent responses (minimax).
 * If any scenario results in the calling team losing a round, caps fails.
 */

const { getFirestore } = require("firebase-admin/firestore");
const { SEATS, PHASE, WRONG_CAPS_PENALTY } = require("./constants");
const {
  cardSuit,
  compareCards,
  getTeamForSeat,
  getPartnerSeat,
  getNextSeat,
  ANTICLOCKWISE_ORDER,
} = require("./deck");
const { generateAllPlayerViews } = require("./views");

const db = () => getFirestore();

/**
 * Check if a player is currently caps-obligated.
 * This runs after every card play for tracking purposes.
 *
 * A player is caps-obligated when there exists ANY ordering of their
 * remaining cards that guarantees winning all remaining rounds,
 * irrespective of how others play.
 *
 * @param {Object} game - Full game state
 * @param {string} seatId - The player to check
 * @returns {boolean} Whether the player can guarantee all remaining rounds
 */
function checkCapsObligation(game, seatId) {
  const play = game.play;
  const trump = game.trump;

  // Can't be caps-obligated if the team has already lost a round
  const myTeam = getTeamForSeat(seatId);
  const hasLostRound = play.completedRounds.some(
    (r) => getTeamForSeat(r.winner) !== myTeam
  );
  if (hasLostRound) return false;

  // Get remaining cards for all players
  const remainingCards = {};
  for (const seat of SEATS) {
    if (game.pccPartnerOut === seat) continue;
    remainingCards[seat] = [...(game.hands[seat] || [])];
  }

  // Get known information from the calling player's perspective
  const knownInfo = deduceKnownInfo(game, seatId);

  // Try all permutations of the caller's remaining cards
  const myCards = remainingCards[seatId];
  const permutations = getPermutations(myCards);

  for (const playOrder of permutations) {
    if (canGuaranteeWithOrder(game, seatId, playOrder, remainingCards, knownInfo)) {
      return true;
    }
  }

  return false;
}

/**
 * Deduce what a player knows about other players' hands
 * from the play history.
 *
 * @returns {Object} { exhaustedSuits: { seat -> Set of suits }, knownCards: { seat -> cards } }
 */
function deduceKnownInfo(game, seatId) {
  const exhaustedSuits = {};
  for (const seat of SEATS) {
    exhaustedSuits[seat] = new Set();
  }

  // Track which suits players have shown they don't have
  for (const round of (game.play?.completedRounds || [])) {
    if (round.cards.length === 0) continue;
    const ledCard = round.cards[0];
    const ledSuit = cardSuit(ledCard.card);

    for (const entry of round.cards) {
      if (entry.seat === round.cards[0].seat) continue; // leader can play anything
      if (entry.card && cardSuit(entry.card) !== ledSuit) {
        // Player didn't follow suit — they're out of this suit
        exhaustedSuits[entry.seat].add(ledSuit);
      }
      if (entry.faceDown && !entry.revealed) {
        // Player played face down — they're out of the led suit
        exhaustedSuits[entry.seat].add(ledSuit);
      }
    }
  }

  return { exhaustedSuits };
}

/**
 * Check if a specific play order guarantees winning all remaining rounds.
 * Uses minimax: for each possible arrangement of opponents' cards,
 * check if the calling team wins all rounds.
 */
function canGuaranteeWithOrder(game, seatId, playOrder, allCards, knownInfo) {
  const myTeam = getTeamForSeat(seatId);
  const partnerSeat = getPartnerSeat(seatId);
  const trumpSuit = game.trump.trumpSuit;
  const trumpRevealed = game.trump.isRevealed || game.trump.isOpen;

  // Build the set of remaining cards for opponents and partner
  const otherSeats = SEATS.filter(
    (s) => s !== seatId && s !== game.pccPartnerOut
  );

  // For each round to be played, simulate
  const remainingRounds = 8 - (game.play?.completedRounds?.length || 0);
  if (playOrder.length < remainingRounds) return false;

  // Get all remaining cards for other players
  const otherCards = {};
  for (const seat of otherSeats) {
    otherCards[seat] = [...(allCards[seat] || [])];
  }

  // Simulate rounds
  const simCards = {
    [seatId]: [...playOrder],
  };
  for (const seat of otherSeats) {
    simCards[seat] = [...otherCards[seat]];
  }

  let currentPriority = game.play?.priority || seatId;

  // For the current round in progress, handle partially played cards
  const currentRoundPlayed = game.play?.currentRound?.length || 0;
  if (currentRoundPlayed > 0) {
    // There's a round in progress — we need to handle this
    // For simplicity, check from the next full round
    // (caps is typically called at the start of a player's turn)
  }

  for (let r = 0; r < remainingRounds; r++) {
    const leader = currentPriority;
    const leaderCard = simCards[leader]?.shift();
    if (!leaderCard) return false;

    const ledSuit = cardSuit(leaderCard);
    let bestCard = leaderCard;
    let bestSeat = leader;
    let bestIsTrump = cardSuit(leaderCard) === trumpSuit;

    // Each other player plays optimally against the calling team
    const turnOrder = getTurnOrder(leader, game.pccPartnerOut);

    for (const seat of turnOrder) {
      if (seat === leader) continue;
      const hand = simCards[seat];
      if (!hand || hand.length === 0) continue;

      // Determine what they can play
      const suitCards = hand.filter((c) => cardSuit(c) === ledSuit);
      const playableCards = suitCards.length > 0 ? suitCards : hand;

      // If this player is on the opposing team, they play to beat us
      // If on our team, they play to help (but we can't rely on choice)
      const isOpponent = getTeamForSeat(seat) !== myTeam;

      if (isOpponent) {
        // Try every possible card — if ANY card beats our current best, they'll play it
        let canWin = false;
        for (const c of playableCards) {
          if (wouldWinAgainst(c, bestCard, bestIsTrump, ledSuit, trumpSuit)) {
            canWin = true;
            break;
          }
        }
        if (canWin) return false; // Opponent can win this round — caps fails

        // They play their worst card (doesn't matter — we need to win regardless)
        hand.splice(hand.indexOf(playableCards[playableCards.length - 1]), 1);
      } else {
        // Partner — play any card (we can't rely on their choice)
        // Remove a card from their hand
        hand.splice(hand.indexOf(playableCards[0]), 1);
      }
    }

    currentPriority = bestSeat;
  }

  return true;
}

/**
 * Check if card A would beat the current best.
 */
function wouldWinAgainst(cardA, currentBest, currentBestIsTrump, ledSuit, trumpSuit) {
  const suitA = cardSuit(cardA);
  const suitBest = cardSuit(currentBest);

  // Trump beats non-trump
  if (suitA === trumpSuit && !currentBestIsTrump) return true;
  if (suitA !== trumpSuit && currentBestIsTrump) return false;

  // Same suit — compare rank
  if (suitA === suitBest) {
    return compareCards(cardA, currentBest) < 0; // lower index = higher power
  }

  // Different non-trump suits — can't beat
  return false;
}

/**
 * Get turn order from a leader, excluding PCC partner.
 */
function getTurnOrder(leader, pccPartnerOut) {
  const order = [leader];
  let current = leader;
  for (let i = 0; i < 3; i++) {
    current = getNextSeat(current);
    if (current !== pccPartnerOut) {
      order.push(current);
    }
  }
  return order;
}

/**
 * Get all permutations of an array (for small arrays only — max ~4 elements).
 */
function getPermutations(arr) {
  if (arr.length <= 1) return [arr];
  const result = [];
  for (let i = 0; i < arr.length; i++) {
    const rest = [...arr.slice(0, i), ...arr.slice(i + 1)];
    const perms = getPermutations(rest);
    for (const perm of perms) {
      result.push([arr[i], ...perm]);
    }
  }
  return result;
}

/**
 * Call Caps — player declares they can win all remaining rounds.
 *
 * @param {Object} data - { code: string, playOrder: string[] }
 * @param {Object} context - Firebase callable context
 */
async function callCaps(data, context) {
  const uid = context.auth?.uid;
  if (!uid) throw new Error("Authentication required.");

  const code = (data.code || "").toUpperCase().trim();
  const playOrder = data.playOrder;
  if (!Array.isArray(playOrder) || playOrder.length === 0) {
    throw new Error("Must specify the play order for remaining cards.");
  }

  const lobbyRef = db().collection("lobbies").doc(code);

  await db().runTransaction(async (transaction) => {
    const lobbyDoc = await transaction.get(lobbyRef);
    if (!lobbyDoc.exists) throw new Error("Game not found.");
    const lobby = lobbyDoc.data();

    const gameRef = lobbyRef.collection("games").doc(lobby.gameId);
    const gameDoc = await transaction.get(gameRef);
    if (!gameDoc.exists) throw new Error("Game not found.");
    const game = gameDoc.data();

    if (game.phase !== PHASE.PLAYING) throw new Error("Not in play phase.");

    const callerSeat = SEATS.find((s) => lobby.seats[s] && lobby.seats[s].uid === uid);
    if (!callerSeat) throw new Error("You are not in this game.");

    const myTeam = getTeamForSeat(callerSeat);
    const trumperTeam = getTeamForSeat(game.trump.trumperSeat);
    const isExternal = myTeam !== trumperTeam;

    // Check if the team has already lost a round
    const hasLostRound = game.play.completedRounds.some(
      (r) => getTeamForSeat(r.winner) !== myTeam
    );
    if (hasLostRound) {
      throw new Error("Cannot call Caps — your team has already lost a round.");
    }

    // Validate the play order matches the caller's remaining cards
    const myHand = game.hands[callerSeat];
    const sortedOrder = [...playOrder].sort();
    const sortedHand = [...myHand].sort();
    if (JSON.stringify(sortedOrder) !== JSON.stringify(sortedHand)) {
      throw new Error("Play order must contain exactly your remaining cards.");
    }

    // Verify the caps claim
    const allCards = {};
    for (const seat of SEATS) {
      if (game.pccPartnerOut === seat) continue;
      allCards[seat] = [...(game.hands[seat] || [])];
    }

    const knownInfo = deduceKnownInfo(game, callerSeat);
    const isValid = canGuaranteeWithOrder(game, callerSeat, playOrder, allCards, knownInfo);

    if (!isValid) {
      // Wrong/Early Caps — 5 stone penalty
      game.play.capsCall = {
        calledBy: callerSeat,
        calledAtRound: game.play.roundNumber,
        playOrder: playOrder,
        type: isExternal ? "external" : "regular",
        result: "wrong_early",
      };

      game.phase = PHASE.COMPLETE;
      game.result = {
        reason: "caps_wrong",
        stoneExchanged: WRONG_CAPS_PENALTY,
        stoneDirection: "receive",
        winnerTeam: myTeam === "teamA" ? "teamB" : "teamA",
        capsBy: callerSeat,
        description: `Wrong/Early Caps by ${callerSeat}. ${WRONG_CAPS_PENALTY} stone penalty.`,
      };

      // Apply penalty
      game.stone[myTeam] += WRONG_CAPS_PENALTY;
    } else {
      // Check timing — was this on time or late?
      const capsObligation = game.play.capsObligations[callerSeat];
      const isLate = capsObligation !== undefined &&
        (capsObligation.obligatedAtRound < game.play.roundNumber ||
         (capsObligation.obligatedAtRound === game.play.roundNumber &&
          capsObligation.obligatedAtCard < game.play.currentRound.length));

      const isBeforeRound7 = game.play.roundNumber < 7;

      let capsResult;
      if (isLate) {
        capsResult = "late";
      } else {
        capsResult = "correct";
      }

      game.play.capsCall = {
        calledBy: callerSeat,
        calledAtRound: game.play.roundNumber,
        playOrder: playOrder,
        type: isExternal ? "external" : "regular",
        result: capsResult,
      };

      // Resolve the game
      game.phase = PHASE.COMPLETE;

      if (capsResult === "correct" && isBeforeRound7) {
        // Correct caps before round 7 — bonus stone
        const { SCORING_TABLE, PCC_SCORING } = require("./constants");
        const bid = game.bidding.highestBid;
        const scoring = game.bidding.isPCC ? PCC_SCORING : SCORING_TABLE[bid];

        if (isExternal) {
          // External caps: betting team loses normal + 1
          const normalLoss = scoring ? scoring.loss : 2;
          game.result = {
            reason: "external_caps",
            stoneExchanged: normalLoss + 1,
            stoneDirection: "receive",
            winnerTeam: myTeam,
            capsBy: callerSeat,
            description: `External Caps (correct, before Round 7). Betting team receives ${normalLoss + 1} stone.`,
          };
          game.stone[trumperTeam] += normalLoss + 1;
        } else {
          // Regular caps: bonus stone on top of normal win
          const normalWin = scoring ? scoring.win : 1;
          game.result = {
            reason: "caps_correct",
            stoneExchanged: normalWin + 1,
            stoneDirection: "give",
            winnerTeam: myTeam,
            capsBy: callerSeat,
            description: `Caps correct (before Round 7). Betting team gives ${normalWin + 1} stone.`,
          };
          game.stone[trumperTeam] = Math.max(0, game.stone[trumperTeam] - (normalWin + 1));
        }
      } else if (capsResult === "correct") {
        // Correct caps after round 7 — no bonus, normal win
        const { SCORING_TABLE, PCC_SCORING } = require("./constants");
        const bid = game.bidding.highestBid;
        const scoring = game.bidding.isPCC ? PCC_SCORING : SCORING_TABLE[bid];

        if (isExternal) {
          const normalLoss = scoring ? scoring.loss : 2;
          game.result = {
            reason: "external_caps",
            stoneExchanged: normalLoss,
            stoneDirection: "receive",
            winnerTeam: myTeam,
            capsBy: callerSeat,
            description: `External Caps (correct, after Round 7). Normal loss applies.`,
          };
          game.stone[trumperTeam] += normalLoss;
        } else {
          const normalWin = scoring ? scoring.win : 1;
          game.result = {
            reason: "caps_correct",
            stoneExchanged: normalWin,
            stoneDirection: "give",
            winnerTeam: myTeam,
            capsBy: callerSeat,
            description: `Caps correct (after Round 7). Normal win applies.`,
          };
          game.stone[trumperTeam] = Math.max(0, game.stone[trumperTeam] - normalWin);
        }
      } else if (capsResult === "late") {
        // Late caps — loss + 1 stone
        const { SCORING_TABLE, PCC_SCORING } = require("./constants");
        const bid = game.bidding.highestBid;
        const scoring = game.bidding.isPCC ? PCC_SCORING : SCORING_TABLE[bid];
        const normalLoss = scoring ? scoring.loss : 2;

        game.result = {
          reason: "caps_late",
          stoneExchanged: normalLoss + 1,
          stoneDirection: "receive",
          winnerTeam: myTeam === "teamA" ? "teamB" : "teamA",
          capsBy: callerSeat,
          description: `Late Caps. ${normalLoss + 1} stone penalty.`,
        };
        game.stone[trumperTeam] += normalLoss + 1;
      }
    }

    transaction.update(gameRef, game);

    const views = generateAllPlayerViews(game, lobby);
    for (const seat of SEATS) {
      if (views[seat]) {
        const viewRef = gameRef.collection("playerViews").doc(seat);
        transaction.set(viewRef, views[seat]);
      }
    }
  });

  return { success: true };
}

module.exports = {
  checkCapsObligation,
  callCaps,
  deduceKnownInfo,
  canGuaranteeWithOrder,
  getPermutations,
};

File: game304/caps.py
"""Caps obligation algorithm for the 304 card game.

Caps is the most complex component of 304. A player is caps-obligated
when there exists *some* committed sequence of plays from their hand
that guarantees winning all remaining rounds, irrespective of how any
other player chooses to play.

The algorithm is full minimax against the rest of the table:
- The caller commits to a fixed play order for their cards.
- Every other player (including the caller's partner) is treated as
  adversarial: for caps to hold, *every* legal sequence of opposing /
  partner plays must result in the caller's team winning every round.
- "Forced plays" (e.g. partner having only one card of the led suit)
  collapse out of the recursion naturally — there is only one branch
  to explore.

Rules:
- Caps must be called at the first opportunity where certainty is
  achieved.
- Correct Caps before Round 7: +1 stone bonus.
- Correct Caps in/after Round 7: normal scoring, no bonus.
- Late Caps: loss + 1 stone penalty (game flipped to loss).
- Wrong/Early Caps: 5 stone penalty.
- External Caps (from opposition): same correctness test, more lenient
  in practice but the algorithm is identical.
"""

from __future__ import annotations

import itertools

from game304.card import Card
from game304.seating import next_seat, partner_seat, team_of
from game304.state import CapsCall, CapsObligation, GameState
from game304.types import Seat, Suit, Team


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def check_caps_obligation(state: GameState, seat: Seat) -> bool:
    """Return ``True`` if ``seat`` is currently caps-obligated.

    A player is caps-obligated when there exists at least one ordering
    of their remaining cards that guarantees winning every remaining
    round, irrespective of how all other players (including their
    partner) might play.

    Used to detect the moment caps obligation first arises (for Late
    Caps detection).
    """
    play = state.play
    if play is None:
        return False

    my_team = team_of(seat)

    # Cannot be obligated if the caller's team has already lost a round
    if any(team_of(r.winner) != my_team for r in play.completed_rounds):
        return False

    my_cards = list(state.hands.get(seat, []))
    if not my_cards:
        return False

    # The caller may not be the leader of the current round if some cards
    # have already been played. Build "to play" plays for in-progress
    # players using simulation; the caller's order is what we vary.
    for play_order in itertools.permutations(my_cards):
        if _caps_holds(state, seat, list(play_order)):
            return True

    return False


def validate_caps_call(
    state: GameState,
    seat: Seat,
    play_order: list[Card],
) -> bool:
    """Return ``True`` if ``play_order`` guarantees all remaining rounds.

    The play order must contain exactly the caller's remaining cards
    (caller is the seat that called caps). Validation assumes the
    declared order is followed strictly; partner and opponents are
    adversarial.
    """
    return _caps_holds(state, seat, list(play_order))


def track_caps_obligation(state: GameState) -> None:
    """Record when each player first became caps-obligated.

    Should be called after every card play and after each round
    resolution. Only seats not already tracked are checked. Each new
    obligation is stamped with the current round number and the number
    of cards already played in the current round at the time of
    detection.

    Per the rules: "Caps cannot be called after the final card of round
    8 is played." Once the game has reached that point the call window
    is closed, so we do not record new obligations there.
    """
    play = state.play
    if play is None:
        return

    expected = 3 if state.pcc_partner_out is not None else 4
    call_window_closed = (
        play.round_number == 8 and len(play.current_round) >= expected
    )
    if call_window_closed:
        return

    for seat in Seat:
        if seat in play.caps_obligations:
            continue
        if state.pcc_partner_out == seat:
            continue
        try:
            if check_caps_obligation(state, seat):
                play.caps_obligations[seat] = CapsObligation(
                    obligated_at_round=play.round_number,
                    obligated_at_card=len(play.current_round),
                )
        except Exception:
            # Tracking is best-effort; never let an obligation check
            # crash the game loop.
            continue


def is_caps_late(state: GameState, seat: Seat) -> bool:
    """Return ``True`` if a caps call by ``seat`` would be late.

    A call is late if obligation was tracked at an earlier point than
    the current state — earlier round, or same round but earlier card
    index.
    """
    play = state.play
    if play is None:
        return False

    obligation = play.caps_obligations.get(seat)
    if obligation is None:
        return False

    if obligation.obligated_at_round < play.round_number:
        return True
    if (
        obligation.obligated_at_round == play.round_number
        and obligation.obligated_at_card < len(play.current_round)
    ):
        return True
    return False


def deduce_exhausted_suits(state: GameState) -> dict[Seat, set[Suit]]:
    """Deduce which suits each player is publicly known to be out of.

    A player is exhausted of a suit if they failed to follow it in a
    completed round (played a different suit face-up, or played a card
    face-down that was either revealed or remained hidden but was
    nonetheless an off-suit play). Used as a heuristic — the simulation
    itself uses actual remaining hands, which is more precise.
    """
    exhausted: dict[Seat, set[Suit]] = {s: set() for s in Seat}
    play = state.play
    if play is None:
        return exhausted

    for r in play.completed_rounds:
        if not r.cards:
            continue
        led_suit: Suit | None = None
        for entry in r.cards:
            if not entry.face_down:
                led_suit = entry.card.suit
                break
        if led_suit is None:
            continue
        for entry in r.cards:
            if entry.card.suit != led_suit:
                exhausted[entry.seat].add(led_suit)
    return exhausted


# ---------------------------------------------------------------------------
# Internal: minimax simulation
# ---------------------------------------------------------------------------


def _caps_holds(
    state: GameState,
    caller_seat: Seat,
    caller_play_order: list[Card],
) -> bool:
    """Return ``True`` if the caller's play order forces an all-rounds win.

    Treats the caller's partner and both opponents as adversarial: caps
    holds iff for *every* legal sequence of plays by the other three
    players, the caller's team wins every remaining round.
    """
    play = state.play
    if play is None:
        return False

    remaining_rounds = 8 - len(play.completed_rounds)
    if remaining_rounds <= 0:
        return True
    if len(caller_play_order) < remaining_rounds:
        return False

    # Build the live simulation hands for everyone (excluding PCC out seat).
    sim_hands: dict[Seat, list[Card]] = {}
    for s in Seat:
        if state.pcc_partner_out == s:
            continue
        sim_hands[s] = list(state.hands.get(s, []))

    # The caller's remaining cards must equal the play_order multiset.
    if sorted(caller_play_order, key=str) != sorted(sim_hands.get(caller_seat, []), key=str):
        return False

    # Mid-round: cards already played in the in-progress round count
    # toward the round outcome but do not affect remaining hands (already
    # removed from sim_hands at this point). Re-attach them.
    in_progress: list[tuple[Seat, Card, bool]] = [
        (e.seat, e.card, e.face_down) for e in play.current_round
    ]

    leader = play.priority if play.priority is not None else caller_seat
    trump_suit = state.trump.trump_suit
    my_team = team_of(caller_seat)

    return _simulate(
        sim_hands=sim_hands,
        caller_seat=caller_seat,
        caller_play_order=list(caller_play_order),
        caller_play_index=0,
        leader=leader,
        in_progress=in_progress,
        remaining_rounds=remaining_rounds,
        trump_suit=trump_suit,
        my_team=my_team,
        pcc_partner_out=state.pcc_partner_out,
    )


def _simulate(
    *,
    sim_hands: dict[Seat, list[Card]],
    caller_seat: Seat,
    caller_play_order: list[Card],
    caller_play_index: int,
    leader: Seat,
    in_progress: list[tuple[Seat, Card, bool]],
    remaining_rounds: int,
    trump_suit: Suit | None,
    my_team: Team,
    pcc_partner_out: Seat | None,
) -> bool:
    """Minimax simulator. Recurses round-by-round, play-by-play."""
    # Determine who has played in the current round and who is next.
    turn_order = _round_turn_order(leader, pcc_partner_out)
    played_seats = [s for s, _, _ in in_progress]

    # Find the next seat to play
    next_idx = len(played_seats)

    if next_idx >= len(turn_order):
        # Round is complete — resolve and recurse.
        winner = _round_winner(in_progress, trump_suit)
        if team_of(winner) != my_team:
            return False
        if remaining_rounds == 1:
            return True
        return _simulate(
            sim_hands=sim_hands,
            caller_seat=caller_seat,
            caller_play_order=caller_play_order,
            caller_play_index=caller_play_index,
            leader=winner,
            in_progress=[],
            remaining_rounds=remaining_rounds - 1,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    next_seat_to_play = turn_order[next_idx]
    led_suit: Suit | None = in_progress[0][1].suit if in_progress else None

    if next_seat_to_play == caller_seat:
        # Caller plays their next committed card.
        if caller_play_index >= len(caller_play_order):
            return False
        card = caller_play_order[caller_play_index]
        hand = sim_hands.get(caller_seat, [])
        if card not in hand:
            return False
        # Must follow suit if able
        if led_suit is not None and any(c.suit == led_suit for c in hand) and card.suit != led_suit:
            return False
        # Apply the play
        new_hands = _hand_remove(sim_hands, caller_seat, card)
        new_in_progress = in_progress + [(caller_seat, card, False)]
        return _simulate(
            sim_hands=new_hands,
            caller_seat=caller_seat,
            caller_play_order=caller_play_order,
            caller_play_index=caller_play_index + 1,
            leader=leader,
            in_progress=new_in_progress,
            remaining_rounds=remaining_rounds,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    # Other player (partner or opponent) — adversarial.
    other_hand = sim_hands.get(next_seat_to_play, [])
    if not other_hand:
        # No cards left — should not happen if remaining_rounds is consistent
        return False

    valid = _valid_plays_for(other_hand, led_suit)
    if not valid:
        return False

    # For caps to hold, every choice this player could make must still
    # result in the caller's team winning all remaining rounds.
    for chosen in valid:
        new_hands = _hand_remove(sim_hands, next_seat_to_play, chosen)
        new_in_progress = in_progress + [(next_seat_to_play, chosen, False)]
        ok = _simulate(
            sim_hands=new_hands,
            caller_seat=caller_seat,
            caller_play_order=caller_play_order,
            caller_play_index=caller_play_index,
            leader=leader,
            in_progress=new_in_progress,
            remaining_rounds=remaining_rounds,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )
        if not ok:
            return False
    return True


def _valid_plays_for(hand: list[Card], led_suit: Suit | None) -> list[Card]:
    """Return all legal cards from ``hand`` given the led suit.

    If the player has any card of the led suit, only those are legal
    (must follow suit). Otherwise any card is legal.
    """
    if led_suit is None:
        return list(hand)
    suited = [c for c in hand if c.suit == led_suit]
    return suited if suited else list(hand)


def _hand_remove(
    sim_hands: dict[Seat, list[Card]], seat: Seat, card: Card
) -> dict[Seat, list[Card]]:
    """Return a copy of ``sim_hands`` with ``card`` removed from ``seat``."""
    new = {s: list(cs) for s, cs in sim_hands.items()}
    new[seat].remove(card)
    return new


def _round_turn_order(
    leader: Seat, pcc_partner_out: Seat | None
) -> list[Seat]:
    """Anticlockwise turn order for one round, skipping the PCC partner."""
    order: list[Seat] = [leader]
    cur = leader
    while len(order) < 4:
        cur = next_seat(cur)
        if cur == leader:
            break
        if cur == pcc_partner_out:
            continue
        order.append(cur)
    return order


def _round_winner(
    plays: list[tuple[Seat, Card, bool]],
    trump_suit: Suit | None,
) -> Seat:
    """Return the winning seat of a completed round.

    The first card played is the led suit. Trump cards beat non-trump.
    Among cards of the same suit (trump or led suit), the one with the
    lowest power index (J=0, 9=1, A=2, ...) wins.
    """
    if not plays:
        raise ValueError("No plays to resolve")

    led_suit = plays[0][1].suit

    # Identify candidates: any trump card outranks any non-trump.
    trumps = [(s, c) for s, c, _ in plays if trump_suit is not None and c.suit == trump_suit]
    if trumps:
        winner = min(trumps, key=lambda sc: sc[1].power)
        return winner[0]
    # No trumps — highest card of led suit wins
    led = [(s, c) for s, c, _ in plays if c.suit == led_suit]
    if not led:
        # Should not happen — the leader played led_suit
        return plays[0][0]
    winner = min(led, key=lambda sc: sc[1].power)
    return winner[0]
